### PR TITLE
DATA-802 | Schema generator and reworked build structure

### PR DESCRIPTION
**Pull request by afiore**
This is a fairly sizeable PR which I am currently targeting to master; we might want to discuss if it's OK for it to be open sourced as it is, or it needs further amends and should target a different branch first.

## General highlights

- [x] Rework build definition to use sbt, writing java classes generated from protobuf in `target/scala-2.13/src_managed`.
- [x] Remove generator commands for data systems that are not currently supported by Lenses (Pulsar, Redis, etc), making a note on how to reinstate this code in the readme.
- [x] Refactor cli parsing logic with the capability of handling multiple commands through the same main class. This involved replacing the old [Scopt](https://github.com/scopt/scopt) library in favour of the more concise [caseapp](https://github.com/alexarchambault/case-app/).
- [x] Implement a command for randomly generating schemas, aimed at facilitating testing the new Explore search functionality.
- [ ] Setup basic CI through Github actions.
- [x] Setup sbt assembly and allow deploying this tool as a standalone, fat jar. 
- [ ] Publish jars to artifactory

## Progress on the schema generator command
- [x] Implement schema generation for Kafka topics (uses AVRO and the Lenses API).
- [x] Implement schema generation for Elastic indices. This is achieved by converting the generated schema in an Elasticsearch type mapping which gets set when creating the index.
- [x] Implement schema generation for Postgres tables.

### CODE REVIEW

File: src/main/scala/io/lenses/data/generator/domain/recursive/CustomerGenerator.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.avro.SchemaBuilder
+import org.apache.avro.generic.GenericData
+import org.apache.avro.generic.GenericRecord
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.apache.kafka.clients.producer.ProducerRecord
+import org.apache.kafka.common.serialization.StringSerializer
+import org.joda.time.DateTime
+
+object CustomerGenerator extends Generator {
+  private val schema = SchemaBuilder
```
Comment (stheppi) [inline src/main/scala/io/lenses/data/generator/domain/recursive/CustomerGenerator.scala:15]: We have avro4s exactly for this. Toavoid  manual schema building and rely on case classes.
Was there a specific reason to have it manual here?

File: src/main/scala/io/lenses/data/generator/http/LensesClient.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            method = Method.POST,
+            uri = baseUrl / "api" / "v1" / "metadata" / "topics"
+          )
+            .withHeaders(
+              Header("X-Kafka-Lenses-Token", auth.value)
+            )
+            .withEntity(
+              Json.obj(
+                "topicName" -> Json.fromString(topicName),
+                "keyType" -> Json.fromString("BYTES"),
```
Comment (stheppi) [inline src/main/scala/io/lenses/data/generator/http/LensesClient.scala:39]: Why Bytes?
We should generate topics with Key being populated

File: src/main/scala/io/lenses/data/generator/schema/DatasetCreator.scala [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  ): IO[Unit]
+}
+
+object DatasetCreator {
+  def Kafka(
+      lensesClient: LensesClient,
+      lensesCreds: LensesCreds
+  ): DatasetCreator =
+    new DatasetCreator {
+      override def create(schemaName: String, schema: Schema)(implicit
```
Comment (stheppi) [inline src/main/scala/io/lenses/data/generator/schema/DatasetCreator.scala:41]: Why not use a T for which we can have a schema for. See Avro4s.

File: src/main/scala/io/lenses/data/generator/schema/Schema.scala [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package io.lenses.data.generator.schema
```
Comment (stheppi) [inline src/main/scala/io/lenses/data/generator/schema/Schema.scala:1]: https://github.com/sksamuel/avro4s 

is there a reason for this AST?

File: src/main/scala/io/lenses/data/generator/schema/DatasetCreator.scala (hunk 3 above)
Comment (afiore) [reply src/main/scala/io/lenses/data/generator/schema/DatasetCreator.scala:41]: Let me clarify: the generator I am working on is specifically intended to produce random schemas (not records) to facilitate testing the Explore search. The approach I have taken is the following:

- define a simple, canonical representations of a schema.
- use scalacheck to automatically generate an arbitrary number of random instances of these canonical schema
- convert the canonical schema into the datasource specific representation (e.g. AVRO, Elastic type mappings, Postgres DLLs).
- Create the schema directly in the source data system (currently, this means creating empty topics/indices/tables).

While this is the main focus of the story I am working on, we will definitely need to implement also other generators that produce random records for a given schema. As you suggest, these might be simply driven by classes (any `T`) and use avro4s.

File: src/main/scala/io/lenses/data/generator/http/LensesClient.scala (hunk 2 above)
Comment (afiore) [reply src/main/scala/io/lenses/data/generator/http/LensesClient.scala:39]: Fair, this is a first pass, but I agree we should also generate a schema for the key

File: src/main/scala/io/lenses/data/generator/domain/recursive/CustomerGenerator.scala (hunk 1 above)
Comment (afiore) [reply src/main/scala/io/lenses/data/generator/domain/recursive/CustomerGenerator.scala:15]: I might refactor this to use avro4s. Haven't touched this class so far, so it should be the same code that we have in the master branch.

**Comment by afiore** [comment]
move here: https://github.com/lensesio-dev/datagen/pull/2

**Pull request closed without merging**